Namespace errors from k8s client when namespace not set

<!--
We use the issue tracker for bugs and feature requests. For general questions and discussion please use http://slack.openwhisk.org/ or https://openwhisk.apache.org/contact.html instead.

Do NOT share passwords, credentials or other confidential information.

Before creating a new issue, please check if there is one already open that
fits the defect you are reporting.
If you open an issue and realize later it is a duplicate of a pre-existing
open issue, please close yours and add a comment to the other.

Issues can be created for either defects or enhancement requests. If you are a committer than please add the labels "bug" or "feature". If you are not a committer please make clear in the comments which one it is, so that committers can add these labels later.

If you are reporting a defect, please edit the issue description to include the
information shown below.

If you are reporting an enhancement request, please include information on what you are trying to achieve and why that enhancement would help you.

For more information about reporting issues, see
https://github.com/apache/openwhisk/blob/master/CONTRIBUTING.md#raising-issues

Use the commands below to provide key information from your environment:
You do not have to include this information if this is a feature request.
-->

## Environment details:

* local deployment - standalone jar
* kubernetes versions:
```
{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```

## Steps to reproduce the issue:

1.   Configure the invoker to use KCF
1.   Create a service user without access to the `--all-namespaces` / `default` namespace via RBAC
1.   Use a kubeconfig that doesn't have a namespace set and uses the service user token

## Provide the expected results and outputs:
Prewarm Containers should start.

## Provide the actual results and outputs:
Prewarm containers don't start and the invoker isn't even listed in: `http://<controller_host_port>/invokers`.

```
[2019-11-04T12:33:26.826Z] [ERROR] Namespace not specified. But operation requires namespace.
io.fabric8.kubernetes.client.KubernetesClientException: Namespace not specified. But operation requires namespace.
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.checkNamespace(OperationSupport.java:172)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleCreate(OperationSupport.java:233)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleCreate(BaseOperation.java:796)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:326)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.create(BaseOperation.java:322)
	at org.apache.openwhisk.core.containerpool.kubernetes.KubernetesClient.run(KubernetesClient.scala:121)
	at org.apache.openwhisk.core.containerpool.kubernetes.KubernetesContainer$.create(KubernetesContainer.scala:72)
	at org.apache.openwhisk.core.containerpool.kubernetes.KubernetesContainerFactory.createContainer(KubernetesContainerFactory.scala:80)
	at org.apache.openwhisk.core.containerpool.ContainerFactory.createContainer(ContainerFactory.scala:100)
	at org.apache.openwhisk.core.containerpool.ContainerFactory.createContainer$(ContainerFactory.scala:92)
	at org.apache.openwhisk.core.containerpool.kubernetes.KubernetesContainerFactory.createContainer(KubernetesContainerFactory.scala:41)
	at org.apache.openwhisk.core.invoker.InvokerReactive.$anonfun$childFactory$2(InvokerReactive.scala:198)
	at org.apache.openwhisk.core.invoker.InvokerReactive.$anonfun$childFactory$2$adapted(InvokerReactive.scala:198)
	at org.apache.openwhisk.core.containerpool.ContainerProxy$$anonfun$1.applyOrElse(ContainerProxy.scala:277)
	at org.apache.openwhisk.core.containerpool.ContainerProxy$$anonfun$1.applyOrElse(ContainerProxy.scala:249)
	at scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:38)
	at akka.actor.FSM.processEvent(FSM.scala:706)
	at akka.actor.FSM.processEvent$(FSM.scala:703)
	at org.apache.openwhisk.core.containerpool.ContainerProxy.processEvent(ContainerProxy.scala:224)
	at akka.actor.FSM.akka$actor$FSM$$processMsg(FSM.scala:700)
	at akka.actor.FSM$$anonfun$receive$1.applyOrElse(FSM.scala:694)
	at akka.actor.Actor.aroundReceive(Actor.scala:539)
	at akka.actor.Actor.aroundReceive$(Actor.scala:537)
	at org.apache.openwhisk.core.containerpool.ContainerProxy.aroundReceive(ContainerProxy.scala:224)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:610)
	at akka.actor.ActorCell.invoke(ActorCell.scala:579)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:268)
	at akka.dispatch.Mailbox.run(Mailbox.scala:229)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:241)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/core/invoker/src/main/resources/application.conf
+++ b/core/invoker/src/main/resources/application.conf
@@ -77,7 +77,6 @@ whisk {
       key: "openwhisk-role"
       value: "invoker"
     }
-
     # Enables forwarding to remote port via a local random port. This mode is mostly useful
     # for development via Standalone mode
     port-forwarding-enabled = false
@@ -87,6 +86,10 @@ whisk {
     #  2. OR yaml formatted multi line string. See multi line config support https://github.com/lightbend/config/blob/master/HOCON.md#multi-line-strings
     #
     #pod-template =
+
+    # Set this optiona string to be the namespace that the invoker should target for adding pods. This allows the invoker to run in a namesapce it doesn't have API access to but add pods to another namespace. See also https://github.com/apache/openwhisk/issues/4711
+    # action-namespace = "ns-actions"
+
   }
 
   # Timeouts for runc commands. Set to "Inf" to disable timeout.

--- a/core/invoker/src/main/resources/application.conf
+++ b/core/invoker/src/main/resources/application.conf
@@ -87,7 +87,9 @@ whisk {
     #
     #pod-template =
 
-    # Set this optiona string to be the namespace that the invoker should target for adding pods. This allows the invoker to run in a namesapce it doesn't have API access to but add pods to another namespace. See also https://github.com/apache/openwhisk/issues/4711
+    # Set this optional string to be the namespace that the invoker should target for adding pods. This allows the invoker to run in a namesapce it doesn't have API access to but add pods to another namespace. See also https://github.com/apache/openwhisk/issues/4711
+    # When not set the underlying client may pickup the namesapce from the kubeconfig or via system property setting.
+    # See https://github.com/fabric8io/kubernetes-client#configuring-the-client for more information.
     # action-namespace = "ns-actions"
 
   }

--- a/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
+++ b/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
@@ -47,7 +47,7 @@ import spray.json._
 import scala.annotation.tailrec
 import scala.collection.mutable
 import scala.concurrent.duration._
-import scala.concurrent.{ExecutionContext, Future, blocking}
+import scala.concurrent.{blocking, ExecutionContext, Future}
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 
@@ -99,7 +99,7 @@ class KubernetesClient(
     .withRequestTimeout(config.timeouts.logs.toMillis.toInt)
   config.actionNamespace match {
     case Some(s) => configBuilder.withNamespace(s)
-    case _ =>
+    case _       =>
   }
   implicit protected val kubeRestClient = new DefaultKubernetesClient(
     configBuilder
@@ -447,7 +447,7 @@ protected[core] final case class TypedLogLine(time: Instant, stream: String, log
 
 protected[core] object TypedLogLine {
 
-  import KubernetesClient.{K8STimestampFormat, parseK8STimestamp}
+  import KubernetesClient.{parseK8STimestamp, K8STimestampFormat}
 
   def readInstant(json: JsValue): Instant = json match {
     case JsString(str) =>

--- a/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
+++ b/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
@@ -94,16 +94,13 @@ class KubernetesClient(
     with ProcessRunner {
   implicit protected val ec = executionContext
   implicit protected val am = ActorMaterializer()
-  protected val configBuilder = new ConfigBuilder()
-    .withConnectionTimeout(config.timeouts.logs.toMillis.toInt)
-    .withRequestTimeout(config.timeouts.logs.toMillis.toInt)
-  config.actionNamespace match {
-    case Some(s) => configBuilder.withNamespace(s)
-    case _       =>
+  implicit protected val kubeRestClient = {
+    val configBuilder = new ConfigBuilder()
+      .withConnectionTimeout(config.timeouts.logs.toMillis.toInt)
+      .withRequestTimeout(config.timeouts.logs.toMillis.toInt)
+    config.actionNamespace.foreach(configBuilder.withNamespace)
+    new DefaultKubernetesClient(configBuilder.build())
   }
-  implicit protected val kubeRestClient = new DefaultKubernetesClient(
-    configBuilder
-      .build())
 
   private val podBuilder = new WhiskPodBuilder(kubeRestClient, config.userPodNodeAffinity, config.podTemplate)
 

--- a/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
+++ b/core/invoker/src/main/scala/org/apache/openwhisk/core/containerpool/kubernetes/KubernetesClient.scala
@@ -47,7 +47,7 @@ import spray.json._
 import scala.annotation.tailrec
 import scala.collection.mutable
 import scala.concurrent.duration._
-import scala.concurrent.{blocking, ExecutionContext, Future}
+import scala.concurrent.{ExecutionContext, Future, blocking}
 import scala.util.control.NonFatal
 import scala.util.{Failure, Success, Try}
 
@@ -76,6 +76,7 @@ case class KubernetesClientConfig(timeouts: KubernetesClientTimeoutConfig,
                                   invokerAgent: KubernetesInvokerAgentConfig,
                                   userPodNodeAffinity: KubernetesInvokerNodeAffinity,
                                   portForwardingEnabled: Boolean,
+                                  actionNamespace: Option[String] = None,
                                   podTemplate: Option[ConfigMapValue] = None)
 
 /**
@@ -93,10 +94,15 @@ class KubernetesClient(
     with ProcessRunner {
   implicit protected val ec = executionContext
   implicit protected val am = ActorMaterializer()
+  protected val configBuilder = new ConfigBuilder()
+    .withConnectionTimeout(config.timeouts.logs.toMillis.toInt)
+    .withRequestTimeout(config.timeouts.logs.toMillis.toInt)
+  config.actionNamespace match {
+    case Some(s) => configBuilder.withNamespace(s)
+    case _ =>
+  }
   implicit protected val kubeRestClient = new DefaultKubernetesClient(
-    new ConfigBuilder()
-      .withConnectionTimeout(config.timeouts.logs.toMillis.toInt)
-      .withRequestTimeout(config.timeouts.logs.toMillis.toInt)
+    configBuilder
       .build())
 
   private val podBuilder = new WhiskPodBuilder(kubeRestClient, config.userPodNodeAffinity, config.podTemplate)
@@ -441,7 +447,7 @@ protected[core] final case class TypedLogLine(time: Instant, stream: String, log
 
 protected[core] object TypedLogLine {
 
-  import KubernetesClient.{parseK8STimestamp, K8STimestampFormat}
+  import KubernetesClient.{K8STimestampFormat, parseK8STimestamp}
 
   def readInstant(json: JsValue): Instant = json match {
     case JsString(str) =>

